Fixes #14097 - Only run DatabaseCleaner truncation on JS tests

dLobatog opened the pull request:
**Problem:**
DatabaseCleaner runs after all integration tests, on truncation mode.
However, for tests where we use rack-test instead of poltergeist, this
is not necessary and slows down the test suite.

**Solution:**
Mark the tests that use javascript and only run DatabaseCleaner on
truncation mode on these, the rest of them can run in transaction mode
(faster). Notice many tests won't have to login twice because of this.

Some basic benchmarking of rake test:integration suggests this is **15% faster**
than before this patch (before, 3m 42s avg, now 3m 09s avg). 

I tried marking tests as `:js => true` using https://github.com/wojtekmach/minitest-metadata , a more elegant approach that would allow us to not login twice for all tests, but the gem doesn't play well with our test_helper. If you have other suggestions I'd love to hear them.

dLobatog committed:
Fixes #14097 - Only run DatabaseCleaner truncation on JS tests

Problem:
DatabaseCleaner runs after all integration tests, on truncation mode.
However, for tests where we use rack-test instead of poltergeist, this
is not necessary and slows down the test suite.

Solution:
Mark the tests that use javascript and only run DatabaseCleaner on
truncation mode on these, the rest of them can run in transaction mode
(faster).

Some basic benchmarking of rake test:integration suggests this is 15% faster
than before this patch (before, 3m 42s avg, now 3m 09s avg).

Additionally this patch allows login to happen only once for integration
tests that require JS (it logged in twice before), and loads fewer
dependencies for integration tests when ran on isolation (rake test:integration)

domcleal commented on test/test_helper.rb:
```diff
@@ -258,9 +258,6 @@ def next_mac(mac)
     end
   end
 
-  # Transactional fixtures do not work with Selenium tests, because Capybara
```
Why remove this, isn't it still accurate?

dLobatog replied:
I thought I changed the line before to 'transaction', I guess I missed it - with that change yeah, it's not accurate

domcleal commented:
Perhaps linked to my inline comment, I think all tests still use truncation and not transaction cleaning with this change.  It looks like you've only changed the driver for the non-JS tests to use the regular rack test driver rather than poltergeist (a good change nonetheless!).

It might be possible to use truncation for all integration tests and transaction for units + functionals though if that's what you intended to do?

I think the speedup comes from limiting the number of host integration tests that use poltergeist.  Other integration tests do something similar, but by putting those tests into a single context and only changing the driver within that context instead of for the whole file.

> I tried marking tests as :js => true using https://github.com/wojtekmach/minitest-metadata , a more elegant approach that would allow us to not login twice for all tests, but the gem doesn't play well with our test_helper. If you have other suggestions I'd love to hear them.

Two thoughts:
1. Split the integration tests into two parent classes, one per driver so the `setup` method that logs in the first time sets the driver first in the PolgergeistIntegrationTest.
2. Override the `setup_admin` method in those tests that change to the JS driver so it changes the driver and then calls `super` to log in.

dLobatog commented:
@domcleal I've changed the default DatabaseCleaner strategy to `transaction`, which I missed on the original PR. The benchmark was done with that change, somehow I discarded it before committing or something.

Splitting the integration test into JS-required vs not-JS files sounds like a good alternative to `minitest-metadata`, I'll do that and ping you again when it's done.

dLobatog commented:
@domcleal I might have overdone this a bit:
- Now the integration tests have their own test_helper. This allows us to load them independently and they would load faster than with test_helper as it doesn't load any of the functional/unit stuff. Obviously removing this dependency should also help a bit with unit/functional
- There's a class that changes the driver to `javascript_driver` and logs in after that. This allows us to not login twice on classes that use poltergeist

Aside from that. the other two things this PR was doing are still there, tests use :truncation instead of :transaction except for Poltergeist tests, and the driver is also changed only when strictly needed.

dLobatog commented:
Updated to remove an empty line detected by Rubocop

domcleal commented on test/integration_test_helper.rb:
```diff
@@ -0,0 +1,123 @@
+ENV["RAILS_ENV"] = "test"
```
Should this file just require `test_helper` to avoid duplication?

dLobatog replied:
I don't think so, the point of the separation is precisely to avoid loading autorun, facet_test_helper, etc... so that the memory footprint and the boot time for these tests is smaller than before.

Since jenkins runs `jenkins:unit` for functionals/units and `jenkins:integration` for integration tests, each of the tasks will load a bit faster & with less memory footprint:
- `jenkins:unit` won't load capybara, integration tests helper, setup poltergeist, database cleaner, etc...
- `jenkins:integration` won't load simplecov, minitest/autorun, basic_rest_response_test, facet_test_helper,  load `foreman_test_helper_additions`

The only 'problem' as you said is that plugins will need to be aware of this change

domcleal replied:
> simplecov, minitest/autorun

I think these are common, should they be in both?

domcleal replied:
> Since jenkins runs jenkins:unit for functionals/units and jenkins:integration for integration tests, each of the tasks will load a bit faster & with less memory footprint:

I'm not sure it'll affect Jenkins all that much as it runs them in a single rake run, but it'll probably help with single test runs for developers.

domcleal replied:
Is minitest/autorun needed here?  I'm unclear about what it's doing in either case to be honest.

dLobatog replied:
Actually it isn't necessary in neither place, it's only necessary to run `MiniTest::Test` tests, bout our tests inherit from `ActiveSupport::TestCase`. I'll remove it entirely.

domcleal commented:
I think it looks good, it's a lot cleaner and as you say, should speed it up in a few ways.

> Now the integration tests have their own test_helper

My only concern about this is that it will break integration tests in plugins, as they'll today require test_helper rather than integration_test_helper.  I see foreman_salt and foreman_docker both have integration tests, so they will be affected.

Apart from reverting the optimisation, it might be good to post to foreman-dev and/or send PRs to affected projects to change the file name.

domcleal commented on test/integration/audit_test.rb:
```diff
@@ -1,4 +1,4 @@
-require 'test_helper'
+require 'integration_test_helper'
 
 class AuditIntegrationTest < ActionDispatch::IntegrationTest
```
This class also has a JS-enabled test, could you either move it or define a second class in this file for it?

dLobatog replied:
fixed

domcleal commented on test/integration/organization_test.rb:
```diff
@@ -1,4 +1,4 @@
-require 'test_helper'
+require 'integration_test_helper'
 
 class OrganizationIntegrationTest < ActionDispatch::IntegrationTest
```
This class also has a JS-enabled test, could you either move it or define a second class in this file for it?

dLobatog replied:
fixed

domcleal commented on test/integration/hostgroup_test.rb:
```diff
@@ -1,4 +1,4 @@
-require 'test_helper'
+require 'integration_test_helper'
 
 class HostgroupIntegrationTest < ActionDispatch::IntegrationTest
```
This class also has a section of JS-enabled tests, could you either move them or define a second class in this file for it?

dLobatog replied:
fixed

domcleal commented on test/integration/subnet_test.rb:
```diff
@@ -1,4 +1,4 @@
-require 'test_helper'
+require 'integration_test_helper'
 
 class SubnetIntegrationTest < ActionDispatch::IntegrationTest
```
This class also has a JS-enabled test, could you either move it or define a second class in this file for it?

dLobatog replied:
fixed

domcleal commented on test/integration/host_js_integration_test.rb:
```diff
@@ -0,0 +1,322 @@
+require 'integration_test_helper'
+require 'integration/shared/host_finders'
+
+class HostJSIntegrationTest < IntegrationTestWithJavascript
```
Minor point on the file naming, though it matches the class name, the others don't use "integration" in the name as they're already in test/integration/ I guess.

domcleal commented on test/integration/host_js_integration_test.rb:
```diff
@@ -0,0 +1,322 @@
+    end
+  end
+end
+
```
Nitpick, empty line.

domcleal commented on test/integration/shared/host_finders.rb:
```diff
@@ -0,0 +1,49 @@
+    page.find('#inherited_puppetclasses_parameters')
+  end
+end
+
```
Nitpick, empty line.

domcleal commented on test/integration/host_test.rb:
```diff
@@ -51,45 +42,6 @@ def class_params
-    end
-
     describe "NIC modal window" do
       setup { skip "Temporarily disabled until issue #9138 gets resolved" }
```
These NIC tests should also probably move since they rely on the JS driver.  Knowing what we know now, we could probably fix them too.

dLobatog commented:
@domcleal Thanks for the review, I:
- Split up classes into js vs non js tests
- Removed empty lines (rubocop should pick that up)
- Enabled previously skipped host form interface tab tests - I had to modify them a bit to work with the latest select2, nothing major though

dLobatog commented:
Last update just changed the names to omit the `Integration` part as it's already implied by the parent class and the directory.

domcleal commented on test/integration/host_js_test.rb:
```diff
@@ -0,0 +1,472 @@
+        modal.find('.interface_domain').select(domain.name)
+
+        modal.has_select?('.interface_subnet', :options => domain.subnets.map(&:to_label))
+        modal.has_select?('.interface_domain', :selected => domain.name)
```
These two has_select? lines appear to be taking a very long time to execute, about 30 seconds each, causing:

```
09:39:26 HostJSTest::NIC modal window::adding interfaces#test_0003_setting primary updates host name = 65.35 s = .
09:40:31 HostJSTest::NIC modal window::adding interfaces#test_0004_selecting domain updates subnet list = 64.95 s = .
```

I'm even seeing it locally, resulting in this stack trace:

<pre>
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/web_socket_server.rb:80:in `call'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/web_socket_server.rb:80:in `recv'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/web_socket_server.rb:80:in `receive'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/web_socket_server.rb:97:in `send'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/server.rb:38:in `send'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/browser.rb:344:in `command'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/browser.rb:67:in `find_within'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/node.rb:17:in `command'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/node.rb:34:in `find'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/poltergeist-1.9.0/lib/capybara/poltergeist/node.rb:38:in `find_xpath'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/capybara-2.6.2/lib/capybara/node/base.rb:107:in `find_xpath'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/capybara-2.6.2/lib/capybara/query.rb:110:in `block in resolve_for'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/capybara-2.6.2/lib/capybara/node/base.rb:80:in `synchronize'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/capybara-2.6.2/lib/capybara/query.rb:106:in `resolve_for'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/capybara-2.6.2/lib/capybara/node/matchers.rb:94:in `block in assert_selector'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/capybara-2.6.2/lib/capybara/node/base.rb:84:in `synchronize'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/capybara-2.6.2/lib/capybara/node/matchers.rb:93:in `assert_selector'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/capybara-2.6.2/lib/capybara/node/matchers.rb:39:in `has_selector?'
/home/dcleal/.rvm/gems/ruby-2.0.0-p353@foreman/gems/capybara-2.6.2/lib/capybara/node/matchers.rb:387:in `has_select?'
test/integration/host_js_test.rb:400:in `block (3 levels) in <class:HostJSTest>'
</pre>

domcleal commented:
To follow on from that inline comment, some of the newly re-enabled tests are quite painfully long when running, which seems to undo the work you've done here:

<pre>
09:38:11 HostJSTest::NIC modal window::editing interfaces#test_0002_ok button updates overview table = 5.02 s = .
09:38:16 HostJSTest::NIC modal window::adding interfaces#test_0001_click on add opens modal = 4.70 s = .
09:38:20 HostJSTest::NIC modal window::adding interfaces#test_0002_ok button adds new interface = 4.63 s = .
09:39:26 HostJSTest::NIC modal window::adding interfaces#test_0003_setting primary updates host name = 65.35 s = .
09:40:31 HostJSTest::NIC modal window::adding interfaces#test_0004_selecting domain updates subnet list = 64.95 s = .
09:41:06 HostJSTest::NIC modal window::switching flags from the overview table#test_0001_switch primary = 35.82 s = .
09:41:42 HostJSTest::NIC modal window::switching flags from the overview table#test_0002_switch provisioning = 35.72 s = .
09:41:48 HostJSTest::NIC modal window::removing interfaces#test_0001_remove interface = 5.68 s = .
</pre>

dLobatog commented:
@domcleal Fixed, a couple of the new tests were using very slow syntax such as `assert !has_selector` (waits for Capybara max time) vs `has_no_selector` (doesn't wait if the element is not there) & slow finders.

dLobatog commented:
For the record, integration tests run at around 3m10s again

dLobatog commented:
Last update removes 'minitest/autorun' from test_helper, since we don't use `MiniTest::Test` or `Minitest::Unit` as the parent class for our tests.

dLobatog commented:
Actually 4 classes, FilterTest, UserTest, FactParserTest and BuildStatusTest were using `stub` (instead of `stubs` which comes from Mocha), which comes from 'minitest/autorun'. So the last update added a require `minitest/mock` on these classes.

domcleal commented:
Merged as 26dade0, thanks @dLobatog!
